js/csv.js

```diff
@@ -0,0 +1,182 @@
+/**
+ * Reads a CSV file and returns a table with the data.
+ *
+ * @param {Blob|File} file The file object to be read.
+ * @param {Function} parser The function to be used in the
+ * CSV parsing operation.
+ */
+export const readCsv = (file, parser = null) => {
+    parser = parser || parseCsvComplex;
+    return new Promise((resolve, reject) => {
+        const reader = new FileReader();
+        reader.onload = e => resolve(parser(e.target.result));
+        reader.onerror = e => reject(e);
+        reader.readAsText(file);
+    });
+};
+
+/**
+ * Creates the contents of the CSV file with the data given
+ * as an array of rows and the headers.
+ *
+ * @param {Array} data Array of rows with cell data.
+ * @param {Array} headers List of headers for the CSV file.
+ * @returns A string containing the information for the CSV.
+ */
+export const buildCsv = (data, headers = []) => {
+    if (Object.keys(data) === 0) return "";
+
+    let csv = headers.toString();
+    for (let i = 0; i < Object.keys(data).length; i++) {
+        let entry = "";
+        for (let j = 0; j < headers.length; j++) {
+            // values containing the ',' character need to be parsed,
+            // otherwise values will not align with fields
+            const value = data[i][headers[j]];
+            if (String(value).includes(",")) {
+                entry += '"' + String(value).replace(/["']/g, "") + '",';
+            } else {
+                // only append truish values
+                entry += value ? value + "," : ",";
+            }

```

Comment:
If there are no headers given, this does not work now. The previous function would return rows of data separated by the delimiter, this one does not if the header is not given. I don't think they worked in all the cases the same way.

Reply:
Totally right, fixed and added tests to the code!